crates/kubelet/src/resources/device_plugin_manager/plugin_connection.rs

```diff
@@ -0,0 +1,320 @@
+    /// Upon updates, this propagates any changes into the `plugins` map and triggers the `NodePatcher` to
+    /// patch the node with the latest values.
+    /// TODO: return error
+    async fn list_and_watch(
+        &self,
+        devices: Arc<RwLock<DeviceMap>>,
+        update_node_status_sender: broadcast::Sender<()>,
+    ) -> anyhow::Result<()> {
+        let mut stream = self
+            .client
+            .clone()
+            .list_and_watch(Request::new(Empty {}))
+            .await?
+            .into_inner();
+        let mut previous_law_devices: HashMap<String, Device> = HashMap::new();
+        while let Some(response) = stream.message().await? {
+            debug!(resource = %self.register_request.resource_name, "Received ListAndWatch message from this resource");
+            if update_devices_map(
+                &self.register_request.resource_name,
+                devices.clone(),
+                &mut previous_law_devices,
+                response,
+            )
+            .await
+            {
+                // TODO handle error -- maybe channel is full
+                update_node_status_sender.send(()).unwrap();
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn allocate(
+        &self,
+        allocate_request: AllocateRequest,
+    ) -> Result<tonic::Response<AllocateResponse>, tonic::Status> {
+        self.client
+            .clone()
+            .allocate(Request::new(allocate_request))
+            .await
+    }
+}
+
+/// This updates the shared device map with the new devices reported by the device plugin.
+/// This iterates through the latest devices, comparing them with the previously reported devices and
+/// updates the shared device map if:
+/// (1) Device modified: DP reporting a previous device with a different health status
+/// (2) Device added: DP reporting a new device
+/// (3) Device removed: DP is no longer advertising a device
+/// If any of the 3 cases occurs, this returns true, signaling that the `NodePatcher` needs to update the
+/// Node status with new devices.
+async fn update_devices_map(
+    resource_name: &str,
+    devices: Arc<RwLock<DeviceMap>>,
+    previous_law_devices: &mut HashMap<String, Device>,
+    response: ListAndWatchResponse,
+) -> bool {
+    let current_devices = response
+        .devices
+        .into_iter()
+        .map(|device| (device.id.clone(), device))
+        .collect::<HashMap<String, Device>>();
+    let mut update_node_status = false;
+
+    for device in current_devices.values() {
+        // (1) Device modified or already registered
+        if let Some(previous_device) = previous_law_devices.get(&device.id) {
+            if previous_device.health != device.health {
+                devices
+                    .write()
+                    .await
+                    .get_mut(resource_name)
+                    .unwrap()
```

Comment:
I'm not sure what you are referring to. the `get_mut` call on the `plugins` hashmap returns an Option<HashMap> which needs to be unwraped in order to access the map and modify it.

Reply:
In the next `else` block, you have this code:

```rust 
match all_devices_map.get_mut(resource_name) {
                Some(resource_devices_map) => {
                    resource_devices_map.insert(device.id.clone(), device.clone());
                }
                None => {
                    let mut resource_devices_map = HashMap::new();
                    resource_devices_map.insert(device.id.clone(), device.clone());
                    all_devices_map.insert(resource_name.to_string(), resource_devices_map);
                }
            }
```

I figure you can do the same thing here to avoid the `unwrap` and possible panic

Reply:
Good point. i think i `unwrap`ed here because if it doesnt exist in the map, things are out of sync.  something should not be removed from the a resources devices map without it also not existing in `previous_law_devices` since all device updates come through `list_and_watch`. Is it still best to handle it?

Reply:
I added the additional matching instead of unwrapping: f96089d1052b306